### PR TITLE
[SYCLomatic] Weekly pulldown from origin/sycl to origin/SYCLomatic 20240715

### PR DESCRIPTION
**Pull request by yihwang-nv**

### CODE REVIEW

File: clang/include/clang/AST/Decl.h [hunk 1]
```diff
@@ -3054,7 +3054,6 @@ class FunctionDecl : public DeclaratorDecl,
   /// computed and stored.
   unsigned getODRHash() const;
 
-<<<<<<< HEAD
 #ifdef SYCLomatic_CUSTOMIZATION
```
Comment (tangjj11) [inline clang/include/clang/AST/Decl.h]: Need handle some github action

File: clang/lib/AST/TypePrinter.cpp [hunk 2]
```diff
@@ -1688,9 +1687,6 @@ void TypePrinter::printTemplateId(const TemplateSpecializationType *T,
   else
 #endif // SYCLomatic_CUSTOMIZATION
   // FIXME: Null TD never excercised in test suite.
```
Comment (tangjj11) [inline clang/lib/AST/TypePrinter.cpp]: Should we apply the change? `// FIXME: Null TD never exercised in test suite.`

**Review by tangjj11 — CHANGES_REQUESTED** [review]

File: clang/lib/Parse/ParseExprCXX.cpp [hunk 3]
```diff
@@ -3030,23 +3029,13 @@ bool Parser::ParseUnqualifiedId(CXXScopeSpec &SS, ParsedType ObjectType,
           SS, ObjectType, ObjectHadErrors,
           TemplateKWLoc ? *TemplateKWLoc : SourceLocation(), Id, IdLoc,
           EnteringContext, Result, TemplateSpecified);
+    else if (TemplateSpecified &&
```
Comment (tangjj11) [inline clang/lib/Parse/ParseExprCXX.cpp]: These revert code do not have `SYCLomatic` tag, and will conflict once llvm or sycl touch these code. Can we just modify the core logic that cause the problem?

**Review by tangjj11 — APPROVED** [review]

**Comment by zhimingwang36** [comment]
LGTM

File: clang/lib/Parse/ParseExprCXX.cpp (hunk 3 above)
Comment (ziranzha) [reply clang/lib/Parse/ParseExprCXX.cpp]: `SYCLomatic_CUSTOMIZATION` is needed here to mark the difference with clang main trunk.

**Review by ziranzha — APPROVED** [review]

**Pull request merged**